src/graphql-types/ReasonApolloQuery.re

```diff
@@ -42,38 +39,41 @@ module QueryFactory = (InternalConfig:InternalConfig) => (ClientConfig: ClientCo
       );
     };
 
-    let component = ReasonReact.reducerComponentWithRetainedProps("ReasonApollo");
-    let make = (~variables=?, children) => {
+    let component = ReasonReact.reducerComponent("ReasonApollo");
+    let make = (~query as q, children) => {
       ...component,
-      initialState: () => Loading,
-      reducer: (action, _state) =>
+      initialState: () => {
+        response: Loading,
+        variables: q##variables
+      },
+      reducer: (action, state) =>
         switch action {
           | Result(result) => {
-            let typedResult = cast(result)##data;
-            ReasonReact.Update(Loaded(typedResult))
+            let typedResult = castResponse(result)##data;
+            ReasonReact.Update({
+              ...state,
+              response: Loaded(typedResult)
+            })
           }
-          | Error(error) => ReasonReact.Update(Failed(error))
+          | Error(error) => ReasonReact.Update({
+            ...state,
+            response: Failed(error)
+          })
         },
-      retainedProps: {variables: variables},
-      willReceiveProps: ({retainedProps, state, reduce}) => {
-        switch (variables, retainedProps.variables) {
-          | (Some(_variables), Some(retainedVariables)) => {
-            if(_variables !== retainedVariables) {
-              sendQuery(~query=ClientConfig.query, ~variables, ~reduce);
-              state;
-            } else {
-              state;
-            }
-          }
-          | _ => state
+      willReceiveProps: ({state, reduce}) => {
+        if(q##variables != state.variables) {
+          sendQuery(~query=q, ~reduce);
+          state;
+        } else {
+          state;
         }
       },
       didMount: ({reduce}) => {
-        sendQuery(~query=ClientConfig.query, ~variables, ~reduce);
+        sendQuery(~query=q, ~reduce);
         ReasonReact.NoUpdate;
       },
       render: ({state}) => {
-        children[0](state);
+        children[0](state.response, q##parse);
```

Comment:
Maybe I'm missing something but I'm wondering why we need to explicitly pass `parse` back to the caller since the caller was the one passing the `query` in the first place 🤔

Also, what's the reason of letting the caller do the parsing instead of doing it internally?

Reply:
> Also, what's the reason of letting the caller do the parsing instead of doing it internally?

1. You want to work on the raw data instead of ReasonML data type
2. Last time I check, you can't parse it internally, the result type is not accessible within ReasonApolloQuery, especially when you parse into a record with `@bsRecord` flag

Reply:
It's not mandatory indeed. Just useful to have it.
I'd like to be able to parse it inside the component and give it back parsed, but as far as I know it's not possible.
Now, the initial state is `Js.Json.t`, but if you want to give the parsed result back you should use 
`external castResult : ...  = "%identity";`, but this declaration is on module level.
Because the type is available in the `make` function, you wont be able to cast it.
Can you confirm please @thangngoc89 ?

Reply:
I see. I guess we can improve it later, don't want to "block" you guys any further 😉

---
Possible approach (just thinking out loud): you pass the query module to the component (not sure it's actually possible) or you pass the `make` function.

```reason
<Query variables={...} gql=PokemonQuery /* <-- is that even possible?? */

<Query variables={...} make=PokemonQuery.make
```

Then you _should_ be able to have control over making the query and parsing it:

```reason
/* Inside QueryFactory module */
let pokemonQuery = make(~name="Pikachu", ());
pokemonQuery.parse(result)
```

But I might be mistaken as well 😅

Reply:
Last time I check, it's not possible. But please make the parsing optional. 

My use case is I have many ReactJS component that consume JS Object so converting the data back and forth is not ideal to me.

Reply:
@emmenko You can't pass a module in a function. Only a functor can take a module as arg, but because a JSX component is a function, it is not possible. Pokemon.make takes labelled arguments, so you will not be able to apply the `variables` record to the make function...

@thangngoc89 good to know some people need optional parsing, so +1 to leave it as is now...